opamp/observiq/observiq_client.go

```diff
@@ -0,0 +1,190 @@
+package observiq
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+
+	"github.com/observiq/observiq-otel-collector/collector"
+	"github.com/observiq/observiq-otel-collector/opamp"
+	"github.com/open-telemetry/opamp-go/client"
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/protobufs"
+	"go.uber.org/zap"
+)
+
+var (
+	// ErrUnsupportedURL is error returned when creating a client with an unsupported URL scheme
+	ErrUnsupportedURL = errors.New("unsupported URL")
+)
+
+// Ensure interface is satisfied
+var _ opamp.Client = (*Client)(nil)
+
+// Client represents a client that is connected to Iris via OpAmp
+type Client struct {
+	opampClient   client.OpAMPClient
+	logger        *zap.SugaredLogger
+	ident         *identity
+	configManager opamp.ConfigManager
+	collector     collector.Collector
+
+	// Parameters needed for connection
+	endpoint  string
+	secretKey string
+}
+
+// NewClientArgs arguments passed when creating a new client
+type NewClientArgs struct {
+	DefaultLogger *zap.SugaredLogger
+	Config        opamp.Config
+	Collector     collector.Collector
+
+	ManagerConfigPath   string
+	CollectorConfigPath string
+	LoggerConfigPath    string
+}
+
+// NewClient creates a new OpAmp client
+func NewClient(args *NewClientArgs) (opamp.Client, error) {
+	clientLogger := args.DefaultLogger.Named("opamp")
+
+	configManager := NewAgentConfigManager(args.DefaultLogger)
+
+	observiqClient := &Client{
+		logger:        clientLogger,
+		ident:         newIdentity(clientLogger, args.Config),
+		configManager: configManager,
+		collector:     args.Collector,
+		endpoint:      args.Config.Endpoint,
+		secretKey:     args.Config.GetSecretKey(),
+	}
+
+	// Parse URL to determin scheme
+	opampURL, err := url.Parse(args.Config.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create collect client based on URL scheme
+	switch opampURL.Scheme {
+	case "ws", "wss":
+		observiqClient.opampClient = client.NewWebSocket(clientLogger)
+	default:
+		return nil, ErrUnsupportedURL
+	}
+
+	// Add configs to config manager
+	configManager.AddConfig(ManagerConfigName, &opamp.ManagedConfig{
+		ConfigPath: args.ManagerConfigPath,
+		Reload:     managerReload(observiqClient, args.ManagerConfigPath),
+	})
+
+	configManager.AddConfig(CollectorConfigName, &opamp.ManagedConfig{
+		ConfigPath: args.CollectorConfigPath,
+		Reload:     collectorReload(observiqClient, args.CollectorConfigPath),
+	})
+
+	configManager.AddConfig(args.LoggerConfigPath, &opamp.ManagedConfig{
+		ConfigPath: args.LoggerConfigPath,
+		Reload:     loggerReload(observiqClient, args.LoggerConfigPath),
+	})
+
+	return observiqClient, nil
+}
+
+// Connect initiates a connection to the OpAmp server
+func (c *Client) Connect(ctx context.Context) error {
+	// Compose and set the agent description
+	if err := c.opampClient.SetAgentDescription(c.ident.ToAgentDescription()); err != nil {
```

Comment:
@cpheps is this here because configs could change now during the opamp lifecycle of the collector?